### DISCUSSION TITLE
callTraitHook with optional route parameter

### DISCUSSION THREAD

**Discussion by DJBellis** [livewire/livewire#7914] [7. Bugs (archived)]
### Livewire version

v3.4.4

### Laravel version

v10.43.0

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

Make a component with `php artisan livewire:make TestComponent`
Make a route that has an optional parameter after a required parameter
```
use App\Livewire\TestComponent;
Route::get('test/component/{required}/{optional?}', TestComponent::class);
```

Add in a trait that has a mount hook

```
<?php

namespace App\Traits;

trait TestTrait
{
    public function mountTestTrait()
    {
    }
}
```

Update the component to use the trait and bind route values via a mount method

```
<?php

namespace App\Livewire;

use App\Traits\TestTrait;
use Illuminate\Support\View;
use Livewire\Component;

class TestComponent extends Component
{
    use TestTrait;
    public $required;
    public $optional;

    public function mount(string $required, ?string $optional = null): void
    {
    }

    public function render(): View
    {
        return view('livewire.test-component');
    }
}
```

Navigate to the route without the optional parameter `{local url}/test/component/required/`

The resulting error is:

```
Cannot use positional argument after named argument during unpacking
```

Which is thrown in 
https://github.com/livewire/livewire/blob/cc164742e74191213db6bee86b04ac9833302d01/src/Features/SupportLifecycleHooks/SupportLifecycleHooks.php#L129

When the trait mount method is called.  This seems to be because
[src/Drawer/ImplicitRouteBinding.php](https://github.com/livewire/livewire/blob/main/src/Drawer/ImplicitRouteBinding.php)

Resolves the default mount parameter required in it's 'named' "required" => {route value} and then the optional variable as a positional 0 => {default value}

From my `dd($params)` before the `wrap($this->component)->$method(...$params);`
```
array:2 [▼ // vendor/livewire/livewire/src/Features/SupportLifecycleHooks/SupportLifecycleHooks.php:135
  "required" => "required"
  0 => null
]
```

So when the values are spread by https://github.com/livewire/livewire/blob/cc164742e74191213db6bee86b04ac9833302d01/src/Features/SupportLifecycleHooks/SupportLifecycleHooks.php#L135

The php error is thrown, because there's a named argument before an optional argument

### Repository or Wirebox URL

https://github.com/DJBellis/traitmountissue

### Additional information

Edited for clarity

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by jals-es** [comment]
@DJBellis any update? I have the same problem.

**Reply by DJBellis** [reply]
@jals-es I haven't had time to do much more digging into this yet, my weekends have been fairly busy since I put this up, but I have some time this weekend when I'll look at it.

**Comment by DJBellis** [comment]
So, this appears to only be an issue in mount when coming from a router, for the standard controller dispatcher this is solved by calling the controller method as `...array_values($params)`, however from a blade injection livewire component this doesn't work as we lose associativity in the component and unless variables are passed in order we wind up getting things like:

`<livewire:testcomponent optional='optional' required='required'>`

For a mount trait method like:

`mountTraitTest(string $required, ?string $optional = null): void`

With resulting values of:
```
$required='optional';
$optional='required';
```

In the standard callHook, which appears to the method is called via:

`wrap($this->component)->__call($name, $params);`

Which uses the ImplicitlyBoundMethod class to parse through the parameters.

Which results in the mount method maintaining proper values of:
```
$required='required';
$optional='optional';
```

As this only appears to be an issue with the router on mounting and I'm not sure if there are reasons that __call() is not called in the callTraitHook method I'm opting away from using that to parse this and propose changing:

https://github.com/livewire/livewire/blob/cc164742e74191213db6bee86b04ac9833302d01/src/Features/SupportLifecycleHooks/SupportLifecycleHooks.php#L21

To:
`$this->callTraitHook('mount', $this->mountParams($params));`

and adding in the function 
```
function mountParams($params = []): array
{
    // Check if parameters passed in have no numeric indexing
    return array_reduce(array_keys($params), fn ($carry, $key) => $carry && ! is_numeric(trim($key)), true)
        // Assume params are associative and pass them as an array
        ? $params
        // Otherwise assume the router has properly run through params and pass them as a dispatcher would
        : array_values($params);
}
```

Looking at the history for SupportLifecycleHooks it appears as though the option was available for __call() to be used in callTraitHook, so I'm assuming that there was a good reason to not use it in that method.  My guess is that the proper target isn't always present in the context and so while it works for a trait it might not work for other injected methods. So attempting to call a wrapped value might result in major breakages, but that's just my guess.

**Comment by jals-es** [comment]
I solve it, removing the optional parameter from de mount() arguments and adding it as class variable.

Try to remove mount() parameters and leave them as class variables above.

```
    public string $required;
    public ?string $optional = null;

    public function mount(): void
    {
    }

```
